### PR TITLE
Fix focus/blur events for light dom

### PR DESCRIPTION
**Pull request by blasten**
This fix is required for https://github.com/PolymerElements/paper-menu/issues/44  
cc @notwaldorf

### CODE REVIEW

File: iron-control-state.html [hunk 1]
```diff
@@ -64,6 +64,24 @@
       this.addEventListener('blur', this._boundFocusBlurHandler, true);
     },
 
+    /**
+     * Returns true if the event was dispatched from a shadow host
+     * @param event Event
+     * @return Boolean
+     */
+    _wasDispatchedFromShadowHost: function(event) {
```
Comment (cdata) [inline iron-control-state.html]: This name is a little misleading. Technically you are testing if an event originated from within the shadow tree or the light tree. "Dispatched from shadow host" suggests to me that the event did or did not originate from the current element (`this`).

File: iron-control-state.html [hunk 2]
```diff
@@ -64,6 +64,24 @@
...
+      var el, path = Polymer.dom(event).path.slice();
+
+      while (path.length > 0 && (el = path.shift())) {
+        if (el === this) {
```
Comment (cdata) [inline iron-control-state.html]: I wonder if this is actually sufficient. Technically, wouldn't `this` be on the path for elements in both the shadow tree and the light tree? It is parent to both subtrees.

File: test/focused-state.html [hunk 3]
```diff
@@ -112,6 +120,35 @@
...
+        input = document.querySelector('#input');
+      });
+
+      test('should not fire the focus event', function() {
+        var nFocusEvents = 0;
+        var nBlurEvents = 0;
+
+        lightDOM.addEventListener('focus', function() {
+          nFocusEvents += 1;
+          MockInteractions.blur(lightDOM.$.input);
```
Comment (cdata) [inline test/focused-state.html]: If this is never supposed to fire, does it technically matter if you blur in here?

File: iron-control-state.html (hunk 2 above)
Comment (blasten) [inline iron-control-state.html]: right. That's why it stops there.

File: iron-control-state.html [hunk 4]
```diff
@@ -72,9 +81,8 @@
       // handled). In either case, we can disregard `event.path`.
 
       if (event.target === this) {
-        var focused = event.type === 'focus';
-        this._setFocused(focused);
-      } else if (!this.shadowRoot) {
+        this._setFocused(event.type === 'focus');
+      } else if (!this.shadowRoot && this._hasOwnerRoot(event)) {
```
Comment (cdata) [inline iron-control-state.html]: `getOwnerRoot()` just tells you that the element is in a root. It doesn't tell you which root. You need to figure out which root, in order to do this correctly. What if the root is `this`?

Comment (blasten) [inline iron-control-state.html]: this is now using https://gist.github.com/blasten/7cffbe034449e3f06375 until it lands in Polymer

**Comment by blasten** [comment]
PTAL @cdata . Now, it's using `isLightDescendant`

File: test/focused-state.html (hunk 3 above)
Comment (blasten) [inline test/focused-state.html]: + fixed

File: iron-control-state.html (hunk 4 above)
Comment (blasten) [inline iron-control-state.html]: \+ fixed

File: iron-control-state.html [hunk 5]
```diff
@@ -72,9 +72,8 @@
       // handled). In either case, we can disregard `event.path`.
 
       if (event.target === this) {
-        var focused = event.type === 'focus';
-        this._setFocused(focused);
-      } else if (!this.shadowRoot) {
+        this._setFocused(event.type === 'focus');
+      } else if (!this.shadowRoot && !this.isLightDescendant(event.target)) {
```
Comment (cdata) [inline iron-control-state.html]: :gem:

**Comment by cdata** [comment]
LGTM :+1:

**Pull request merged**